### ISSUE TITLE
Aspire does not run on a distroless image (e.g. CBL Mariner)

### ISSUE THREAD

**Issue by mdrakiburrahman** [dotnet/aspire#5334]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Describe the bug

Here's our dockerfile for production. We want to run on a minimal, distroless image for security (production must run distroless, non-root images) and agility purposes (image should be small and fast to spin up).

```
# ------------------------------------------------------------------------------
# BUILD CONTAINER: FULL DOTNET SDK
# ------------------------------------------------------------------------------
FROM mcr.microsoft.com/dotnet/sdk:8.0 AS build

ARG BUILD_CONFIGURATION=Release
ARG PORT=8080
ARG PROJ_PATH="./src/A/B/C.csproj"
ARG VSS_NUGET_ACCESSTOKEN

RUN curl -L https://raw.githubusercontent.com/Microsoft/artifacts-credprovider/master/helpers/installcredprovider.sh  | sh
ENV VSS_NUGET_EXTERNAL_FEED_ENDPOINTS="{\"endpointCredentials\": [{\"endpoint\":\"https://msdata.pkgs.visualstudio.com/tina/_packaging/adshybrid-Consumption/nuget/v3/index.json\", \"password\":\"${VSS_NUGET_ACCESSTOKEN}\"}]}"

COPY . .
RUN dotnet restore $PROJ_PATH
RUN dotnet build $PROJ_PATH -c $BUILD_CONFIGURATION -o /app/build

# ------------------------------------------------------------------------------
# PUBLISH CONTAINER: SAME AS BUILD
# ------------------------------------------------------------------------------
FROM build AS publish

RUN dotnet publish $PROJ_PATH --no-restore -c $BUILD_CONFIGURATION -o /app/publish

# ------------------------------------------------------------------------------
# RUNTIME CONTAINER: DISTROLESS (I.E. NO SHELL, TDNF PACKAGES ETC.)
# ------------------------------------------------------------------------------
FROM mcr.microsoft.com/dotnet/aspnet:8.0-cbl-mariner2.0-distroless@sha256:c239a8377190f792c245720fae85825c26be351815b7178d9ecb5e6384081ce1 AS final

WORKDIR /app
COPY --from=publish /app/publish .

# ------------------------------------------------------------------------------
# ENTRYPOINT: OVERRIDE VIA RUNTIME INFRA - AKS, ACA, DOCKER DESKTOP ETC.
# ------------------------------------------------------------------------------
EXPOSE $PORT
ENTRYPOINT ["/usr/bin/dotnet", "/app/your-runtime-dll-name.dll"]
```

I was able to get Aspire to build inside this image.

But, at runtime, it throws, since it expects aspire to be installed at runtime:

```
PS C:\Users\mdrrahman> docker run -e "ASPIRE_ALLOW_UNSECURED_TRANSPORT=true" `
>>    -e "ASPNETCORE_ENVIRONMENT=Development" `
>>    -e "DOTNET_ENVIRONMENT=Development" `
>>    -e "DOTNET_DASHBOARD_OTLP_ENDPOINT_URL=http://localhost:21146" `
>>    -e "DOTNET_RESOURCE_SERVICE_ENDPOINT_URL=http://localhost:22234" `
>>    -e "ASPNETCORE_URLS=http://localhost:5000" `
>>    --entrypoint "/usr/bin/dotnet" `
>>    -it dataplane-event-processing-service "/app/Dataplane.EventProcessing.AppHost.dll"
info: Aspire.Hosting.DistributedApplication[0]
      Aspire version: 8.1.0+d304c5f6f15bcd4f34f1841b33870cfab88e6937
info: Aspire.Hosting.DistributedApplication[0]
      Distributed application starting.
info: Aspire.Hosting.DistributedApplication[0]
      Application host directory is: /src/Microsoft.Universal.Aspire.AppHost/Dataplane.EventProcessing/Dataplane.EventProcessing.AppHost
warn: Microsoft.AspNetCore.DataProtection.Repositories.FileSystemXmlRepository[60]
      Storing keys in a directory '/home/app/.aspnet/DataProtection-Keys' that may not be persisted outside of the container. Protected data will be unavailable when container is destroyed. For more information go to https://aka.ms/aspnet/dataprotectionwarning
warn: Microsoft.AspNetCore.DataProtection.KeyManagement.XmlKeyManager[35]
      No XML encryptor configured. Key {b5ed2af9-be9b-418b-86e5-9fab6fd69f2f} may be persisted to storage in unencrypted form.
fail: Microsoft.Extensions.Hosting.Internal.Host[11]
      Hosting failed to start
      System.IO.FileNotFoundException: The Aspire orchestration component is not installed at "/usr/share/dotnet/packs/Aspire.Hosting.Orchestration.linux-x64/8.1.0/tools/dcp". The application cannot be run without it.
      File name: '/usr/share/dotnet/packs/Aspire.Hosting.Orchestration.linux-x64/8.1.0/tools/dcp'
         at Aspire.Hosting.Dcp.DcpDependencyCheck.GetDcpInfoAsync(CancellationToken cancellationToken) in /_/src/Aspire.Hosting/Dcp/DcpDependencyCheck.cs:line 50
         at Aspire.Hosting.Dcp.DcpHostService.StartAsync(CancellationToken cancellationToken) in /_/src/Aspire.Hosting/Dcp/DcpHostService.cs:line 67
         at Microsoft.Extensions.Hosting.Internal.Host.<StartAsync>b__15_1(IHostedService service, CancellationToken token)
         at Microsoft.Extensions.Hosting.Internal.Host.ForeachService[T](IEnumerable`1 services, CancellationToken token, Boolean concurrent, Boolean abortOnFirstException, List`1 exceptions, Func`3 operation)
Unhandled exception. System.AggregateException: One or more errors occurred. (The Aspire orchestration component is not installed at "/usr/share/dotnet/packs/Aspire.Hosting.Orchestration.linux-x64/8.1.0/tools/dcp". The application cannot be run without it.)
 ---> System.IO.FileNotFoundException: The Aspire orchestration component is not installed at "/usr/share/dotnet/packs/Aspire.Hosting.Orchestration.linux-x64/8.1.0/tools/dcp". The application cannot be run without it.
File name: '/usr/share/dotnet/packs/Aspire.Hosting.Orchestration.linux-x64/8.1.0/tools/dcp'
...
```

I can't install aspire in a distroless container either:

```
[+] Building 1.6s (8/15)                                                                   docker:desktop-linux
 => [internal] load build definition from Dockerfile.csharp.runtime                                        0.0s
 => => transferring dockerfile: 2.02kB                                                                     0.0s
 => [internal] load metadata for mcr.microsoft.com/dotnet/aspnet:8.0-cbl-mariner2.0-distroless@sha256:c23  0.2s
 => [internal] load metadata for mcr.microsoft.com/dotnet/sdk:8.0                                          0.0s
 => [internal] load .dockerignore                                                                          0.0s
 => => transferring context: 2B                                                                            0.0s
 => [build 1/6] FROM mcr.microsoft.com/dotnet/sdk:8.0                                                      0.0s
 => CACHED [final 1/4] FROM mcr.microsoft.com/dotnet/aspnet:8.0-cbl-mariner2.0-distroless@sha256:c239a837  0.0s
 => CANCELED [internal] load build context                                                                 0.4s
 => => transferring context: 821.70kB                                                                      0.4s
 => ERROR [final 2/4] RUN dotnet workload install aspire                                                   0.4s
------
 > [final 2/4] RUN dotnet workload install aspire:
0.341 runc run failed: unable to start container process: exec: "/bin/sh": stat /bin/sh: no such file or directory
------
Dockerfile.csharp.runtime:31
--------------------
  29 |     FROM mcr.microsoft.com/dotnet/aspnet:8.0-cbl-mariner2.0-distroless@sha256:c239a8377190f792c245720fae85825c26be351815b7178d9ecb5e6384081ce1 AS final
  30 |
  31 | >>> RUN dotnet workload install aspire
  32 |     WORKDIR /app
  33 |     COPY --from=build /app/build .
--------------------
ERROR: failed to solve: process "/bin/sh -c dotnet workload install aspire" did not complete successfully: exit code: 1
PS C:\Users\mdrrahman\Documents\GitHub\AdsHybrid>
```

Since distroless containers do not have shells.

**Comment by mdrakiburrahman** [comment]
I see from the `aspire-dashboard` Dockerfile, that the build is downloaded from a Zipfile from the internet and copied into the distroless image:

https://github.com/dotnet/dotnet-docker/blob/main/src/aspire-dashboard/8.1/cbl-mariner-distroless/amd64/Dockerfile

This becomes difficult to do during regular CICD - it would be amazing if there was a self-contained Dockerfile that showed a pattern how to run Aspire (Dashboard + host) inside a distroless image

**Comment by mdrakiburrahman** [comment]
I was able to use the aspire dashboard Docker image as inspiration to get regular Aspire building + running in a plain old Distroless Dockerfile.

The trick is to copy the aspire packs into the distroless container:

```Dockerfile
# ------------------------------------------------------------------------------
# BUILD CONTAINER: FULL DOTNET SDK
# ------------------------------------------------------------------------------
FROM mcr.microsoft.com/dotnet/sdk:8.0 AS build

ARG BUILD_CONFIGURATION=Release
ARG PORT=8080
ARG PROJ_PATH="./src/A/B/C.csproj"
ARG VSS_NUGET_ACCESSTOKEN

RUN curl -L https://raw.githubusercontent.com/Microsoft/artifacts-credprovider/master/helpers/installcredprovider.sh  | sh
ENV VSS_NUGET_EXTERNAL_FEED_ENDPOINTS="{\"endpointCredentials\": [{\"endpoint\":\"https://msdata.pkgs.visualstudio.com/tina/_packaging/adshybrid-Consumption/nuget/v3/index.json\", \"password\":\"${VSS_NUGET_ACCESSTOKEN}\"}]}"

COPY . .
RUN dotnet workload install aspire
RUN dotnet restore $PROJ_PATH
RUN dotnet build $PROJ_PATH -c $BUILD_CONFIGURATION -o /app/build

# ------------------------------------------------------------------------------
# RUNTIME CONTAINER: DISTROLESS (I.E. NO SHELL, TDNF PACKAGES ETC.)
# ------------------------------------------------------------------------------
FROM mcr.microsoft.com/dotnet/aspnet:8.0.8-cbl-mariner2.0-distroless-extra-amd64@sha256:3818d16610477e3a8b19da5e6e94a4de279ed79cb55f5580e1797d29b97ab304 AS final

WORKDIR /usr/share/dotnet/packs
COPY --from=build /usr/share/dotnet/packs .

WORKDIR /app
COPY --from=build /app/build .

# ------------------------------------------------------------------------------
# ENTRYPOINT: OVERRIDE VIA RUNTIME INFRA - AKS, ACA, DOCKER DESKTOP ETC.
# ------------------------------------------------------------------------------
EXPOSE $PORT
ENTRYPOINT ["/usr/bin/dotnet", "/app/your-runtime-dll-name.dll"]
```

And here's the csproj for the App Host:

```xml
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <OutputType>Exe</OutputType>
    <CodeAnalysisRuleSet>..\..\CodeAnalysis.ruleset</CodeAnalysisRuleSet>
    <GenerateDocumentationFile>True</GenerateDocumentationFile>
    <ImplicitUsings>enable</ImplicitUsings>
    <IsAspireHost>true</IsAspireHost>
    <Nullable>enable</Nullable>
    <TargetFramework>net8.0</TargetFramework>
    <UserSecretsId>6c4b049a-7eba-4d98-8a44-503e3c6c54e8</UserSecretsId>
  </PropertyGroup>

  <ItemGroup>
    <PackageReference Include="Aspire.Hosting.AppHost" />
    <PackageReference Include="StyleCop.Analyzers">
      <PrivateAssets>all</PrivateAssets>
      <IncludeAssets>runtime; build; native; contentfiles; analyzers; buildtransitive</IncludeAssets>
    </PackageReference>
  </ItemGroup>

  <ItemGroup>
    <AdditionalFiles Include="..\..\.editorconfig" Link=".editorconfig" />
    <AdditionalFiles Include="..\..\stylecop.json" Link="stylecop.json" />
  </ItemGroup>

  <ItemGroup>
    <ProjectReference Include="..\Your.Service.One\Your.Service.One.csproj" />
    <ProjectReference Include="..\Your.Service.Two\Your.Service.Two.csproj" />
  </ItemGroup>

  <ItemGroup>
    <None Update="appsettings.json">
      <CopyToOutputDirectory>Always</CopyToOutputDirectory>
    </None>
  </ItemGroup>

</Project>

```

Here's how I can run it with Docker Desktop:

```powershell
docker run -e "ASPIRE_ALLOW_UNSECURED_TRANSPORT=true" `
		   -e "ASPNETCORE_URLS=http://0.0.0.0:18888" `
		   -e "DOTNET_DASHBOARD_UNSECURED_ALLOW_ANONYMOUS=true" `
		   -e "DOTNET_DASHBOARD_OTLP_ENDPOINT_URL=http://0.0.0.0:18889" `
		   -e "DOTNET_DASHBOARD_OTLP_HTTP_ENDPOINT_URL=http://0.0.0.0:18890" `
		   -p 18888:18888 `
		   --entrypoint "/usr/bin/dotnet" `
		   -it dataplane-event-processing-service "/app/Your.Service.AppHost.dll"
```

![image](https://github.com/user-attachments/assets/28268261-0203-484a-b368-e5b86c06c888)

**Comment by mdrakiburrahman** [comment]
See above

**Comment by mdrakiburrahman** [comment]
(I reopened the issue because Aspire threw at runtime)

**Comment by mdrakiburrahman** [comment]
Hey @davidfowl -

We're ultimately looking to host Aspire on Kubernetes (AKS, and local Microk8s clusters for devs), and also hosting in Docker Desktop for CI tests etc.

Finally, Microsoft AME tenant deployments must use a Azure Container Registry to deploy to production regions.

We cannot use `azd` for our AKS deployments, so the easiest way is to have a Container Image.

There's not a lot of documentation available on the Aspire docs for AKS, besides Azure Container Apps.

I found this [Aspirate CLI tool](https://github.com/prom3theu5/aspirational-manifests) some person wrote, that apparently converts Aspire manifests to Kubernetes YAML, but my team cannot take a dependency on this CLI tool (because we can't use it to deploy to Microsoft AME tenant).

Besides containerizing the Aspire `AppHost` project - do you have any guidance on the best way to deploy an Aspire hosted app that wraps many C# microservices into Production on AKS?

**Comment by davidfowl** [comment]
Stepping back a bit spending some time in [the docs](https://learn.microsoft.com/en-us/dotnet/aspire/get-started/aspire-overview) will make some of what is possible today clearer. 

To answer some of your questions directly:

> Besides containerizing the Aspire AppHost project - do you have any guidance on the best way to deploy an Aspire hosted app that wraps many C# microservices into Production on AKS?

The apphost project is used locally (and for testing) and is never deployed. The only first class experience for deploying aspire apps to AKS is aspir8 (the tool you found). To understand more about how it works, see:

https://learn.microsoft.com/en-us/dotnet/aspire/deployment/manifest-format

TL;DR it can analyze the aspire manifest and emit kube manifests (based on kustomize) or helm charts (as well as pushing container images).

> Is it a bad idea to containerize AppHost and run it (e.g. at least for local/CI testing)?

I'm not sure if it's a bad idea but what's the goal of running the apphost project in a container? What are looking to accomplish?

> Also, there's not a lot of documentation/tutorials available on the Aspire docs for AKS/plain-old-kubernetes, besides Azure Container Apps - which uses azd.

PS: If you're trying to go to AKS then azd provides nothing for you today anyways. It's on the roadmap but we're focusing on finishing a single compute platform (ACA) until fan out to more. 

> We're ultimately looking to host Aspire on Kubernetes (AKS, and local Microk8s clusters for devs), and also hosting in Docker Desktop/Docker compose for CI tests and rapid regression testing etc. All of our code runtime is Linux, and we have standardized, hardened containers for production - we're trying to introduce Aspire in this setup.

This is a nit but it's important to clarify that you are not "running aspire on AKS", that gives the impression that the apphost runs on kuberentes and it doesn;'t.
